src/aotcompile.cpp

```diff
@@ -792,6 +615,157 @@ void *jl_create_native_impl(LLVMOrcThreadSafeModuleRef llvmmod, int trim, int ex
     return data;
 }
 
+// x16, x17 (ip0, ip1) are the intra-procedure-call scratch registers
+const char *plt_asm_aarch64_macho =
+    "adrp  x16, ${0}@page\n"
+    "ldr   x16, [x16, ${0}@pageoff]\n"
+    "br    x16\n";
+
+const char *plt_asm_aarch64 =
+    "adrp  x16, ${0}\n"
+    "ldr   x16, [x16, :lo12:${0}]\n"
+    "br    x16\n";
+
+const char *plt_asm_x86_64 =
+    "jmpq  *${0:a}\n";
+
+const char *plt_asm_riscv64 =
+    "1b: auipc t3, %pcrel_hi(${0})\n"
+    "    ld    t3, %pcrel_lo(1b)(t3)\n"
+    "    jalr  t1, t3";
+
+const char *plt_asm_riscv32 =
+    "1b: auipc t3, %pcrel_hi(${0})\n"
+    "    lw    t3, %pcrel_lo(1b)(t3)\n"
+    "    jalr  t1, t3";
+
+// Emit a thunk that calls a compiled CodeInstance from an external image.
+static Function *emit_pkg_plt_thunk(jl_codegen_output_t &out, jl_code_instance_t *ci,
+                                    Function *CallSite)
+{
+    auto &M = out.get_module();
+    auto &Ctx = out.get_context();
+    Type *PtrTy = PointerType::getUnqual(Ctx);
+    StringRef Name = name_from_method_instance(jl_get_ci_mi(ci));
+    auto GVName = out.make_name(JL_SYM_JLPLT_GOT, Name);
+
+    auto GV = new GlobalVariable(M, PtrTy, false, GlobalVariable::ExternalLinkage, nullptr,
+                                 GVName);
+
+
+    const char *Code = nullptr;
+    auto OF = out.TargetTriple.getObjectFormat();
+    if (out.TargetTriple.isAArch64()) {
+        if (OF == Triple::MachO)
+            Code = plt_asm_aarch64_macho;
+        else
+            Code = plt_asm_aarch64;
+    }
+    else if (out.TargetTriple.getArch() == Triple::x86_64) {
+        Code = plt_asm_x86_64;
+    }
+    else if (out.TargetTriple.getArch() == Triple::riscv64) {
+        Code = plt_asm_riscv64;
+    }
+    else if (out.TargetTriple.getArch() == Triple::riscv32) {
+        Code = plt_asm_riscv32;
+    }
+
+    auto FTy = FunctionType::get(Type::getVoidTy(Ctx), !Code);
+    auto F = Function::Create(FTy, Function::PrivateLinkage, 0,
+                              out.make_name(JL_SYM_JLPLT, Name), &M);
+    F->setCallingConv(CallSite->getCallingConv());
+    AttrBuilder Attrs{Ctx};
+    Attrs.addAttribute(Attribute::NoInline);
+    Attrs.addAttribute(Attribute::NoUnwind);
+    Attrs.addAttribute("frame-pointer", "none");
+    Attrs.addAttribute("thunk");
+
+    IRBuilder<> B{Ctx};
+    auto BB = BasicBlock::Create(Ctx, "", F);
+    B.SetInsertPoint(BB);
+
+    if (Code) {
+        Attrs.addAttribute(Attribute::Naked);
+        auto AsmTy = FunctionType::get(Type::getVoidTy(Ctx), {PtrTy}, false);
+        auto Call = B.CreateCall(InlineAsm::get(AsmTy, Code, "s", true, false), {GV});
+        Call->addFnAttr(Attribute::NoReturn);
+        B.CreateUnreachable();
+    }
+    else {
+        // Generic fallback that may be inefficient but won't mangle registers.
+        auto FPtr = B.CreateAlignedLoad(PtrTy, GV, out.DL.getPointerABIAlignment(0));
+        auto Call = B.CreateCall(FTy, FPtr, {});
```

Comment:
you probably do need to pass along the actual arguments to keep llvm happy

Reply:
I believe it's special-cased in the verifier:
> If the musttail call appears in a function with the "thunk" attribute and the caller and callee both have varargs, then any unprototyped arguments in register or memory are forwarded to the callee. Similarly, the return value of the callee is returned to the caller’s caller, even if a void return type is in use.